### ISSUE TITLE
Multi Lora support using "fuse_lora"

### ISSUE THREAD

**Issue by adhikjoshi** [huggingface/diffusers#4919]
I have experimented with the "fuse_lora" function to check if multi-lora can be supported, and it works.

I can see all 3 Lora influencing results.

```
from diffusers import DiffusionPipeline
import torch 

pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16).to("cuda")

#Lora 1
# pipe.load_lora_weights(".", weight_name="lora_models/dd.safetensors")
# pipe.fuse_lora()

# Lora 2
#pipe.load_lora_weights(".", weight_name="lora_models/pxart.safetensors")
#pipe.fuse_lora()

# Lora 3
pipe.load_lora_weights(".", weight_name="lora_models/ink.safetensors")
pipe.fuse_lora()


generator = torch.manual_seed(0)
images_fusion = pipe("masterpiece, best quality, mountain", generator=generator, num_inference_steps=20).images[0]
images_fusion.save("out/img8.png")
```

There is a warning

> "The current API is supported for operating with a single LoRA file. You are trying to load and fuse more than one LoRA which is not well-supported."

If we can get native support for Multi Lora @sayakpaul @patrickvonplaten using existing API

**Comment by sayakpaul** [comment]
Cool that you tried that out. Possible to share some results here?

**Comment by patrickvonplaten** [comment]
Yes, multi-lora is definitely on the roadmap

**Comment by Damarcreative** [comment]
no wonder I can't use more than one lora
thank you, I will try it,

**Comment by landmann** [comment]
Hey guys! Has multi-lora been supported? Would be super useful!

**Comment by Bec-k** [comment]
Is this implemented?

**Comment by octa9999** [comment]
I want to use multiple loras like, i can load 3 loras in the start and i can use whichever i want or multiple by using trigger keywords, how can i achieve this?

**Comment by sayakpaul** [comment]
Have you gone through https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference ?

**Comment by octa9999** [comment]
yes, I have
by using this method, I have to always call set_adapter to activate a particular lora, what i want to achieve is that i load all the loras before inference and i don't have to set the particular lora, i can just use the trigger word and get output results influenced by the lora instead of manually setting it everytime, can i achieve this functionality using diffusers ?

**Comment by sayakpaul** [comment]
You can call `load_lora_weights()` on all the adapters.

Then before running inference, you do `set_adapter()` on the adapters you loaded.

**Comment by octa9999** [comment]
can i set all the adapters at once, for example i load 3 different loras, and after calling load_lora_weights() three times i call set_adapter([1.0, 1.0, 1.0]) and i can just use trigger words to trigger the particular lora or multiple loras?

Thank you for your help

**Comment by octa9999** [comment]
does this work in parallel / concurrent inferencing, lets say there are 4 different threads inferencing at the same time, and all 4 threads are using different lora adapters while using same base model ? 
Because all 4 threads will be setting different adapters, please guide me about this, thank you

**Comment by asrlhhh** [comment]
Hi has anyone tried saving the fused lora model and later reloading it. When I tried, I am facing errors of mismatched weight params

```
pipe_id = "stabilityai/stable-diffusion-xl-base-1.0"
pipe = DiffusionPipeline.from_pretrained(pipe_id, torch_dtype=torch.float16)#.to("cuda")
pipe.load_lora_weights("sdxl-lora/pytorch_lora_weights.safetensors")
pipe.fuse_lora()
pipe.save_pretrained("sdxl-lora-fused")
```

```
pipe_id = "sdxl-lora-fused"
pipe = DiffusionPipeline.from_pretrained(pipe_id, torch_dtype=torch.float16).to("cuda")
```

The errors:

```
ValueError: Cannot load <class 'diffusers.models.unet_2d_condition.UNet2DConditionModel'> from sdxl-lora-cognex-fused4/unet because the following keys are missing: 
 mid_block.attentions.0.transformer_blocks.5.attn2.to_q.weight, down_blocks.2.attentions.1.transformer_blocks.7.attn1.to_v.weight, up_blocks.0.attentions.2.transformer_blocks.7.attn1.to_q.weight, up_blocks.0.attentions.2.transformer_blocks.8.attn2.to_q.weight, mid_block.attentions.0.transformer_blocks.0.attn1.to_out.0.weight, up_blocks.0.attentions.1.transformer_blocks.6.attn2.to_q.weight, mid_block.attentions.0.transformer_blocks.2.attn1.to_q.weight, up_blocks.0.attentions.0.transformer_blocks.0.attn1.to_out.0.bias, down_blocks.2.attentions.0.transformer_blocks.1.attn2.to_q.weight
 ....
 mid_block.attentions.0.transformer_blocks.1.attn1.to_out.0.bias, mid_block.attentions.0.transformer_blocks.5.attn1.to_out.0.bias, down_blocks.2.attentions.1.transformer_blocks.5.attn1.to_out.0.weight, down_blocks.2.attentions.0.transformer_blocks.1.attn1.to_q.weight, down_blocks.2.attentions.0.transformer_blocks.1.attn2.to_v.weight, up_blocks.0.attentions.1.transformer_blocks.3.attn1.to_out.0.weight. 
 Please make sure to pass `low_cpu_mem_usage=False` and `device_map=None` if you want to randomly initialize those weights or else make sure your checkpoint file is correct.
```

**Comment by asrlhhh** [comment]
Thanks for the pointer. You mentioned that without peft it works. So how do I turn off peft in the above code?

**Comment by sayakpaul** [comment]
You uninstall it. But I have not checked it for multi Adapters.

**Comment by asrlhhh** [comment]
yeah uninstalled it and it works. However I still need the peft version to work because the DPO script is full of peft. Changing all the places to non peft versions would be too much work

**Comment by sayakpaul** [comment]
Yeah needs to wait for the bug to be fixed.

**Comment by jslegers** [comment]
I was trying to do this :

```python
pipe.load_lora_weights(lora1_repo, weight_name=lora1_filename, adapter_name='lora1')
pipe.fuse_lora(lora_scale=0.33)
pipe.unload_lora_weights()
pipe.load_lora_weights(lora2_repo, weight_name=lora2_filename, adapter_name='lora2')
pipe.fuse_lora(lora_scale=0.32)
pipe.unload_lora_weights()
pipe.load_lora_weights(lora3_repo, weight_name=lora3_filename, adapter_name='lora3')
pipe.fuse_lora(lora_scale=0.29)
pipe.unload_lora_weights()
pipe.save_pretrained("/content/newModelTest", variant = "fp16")
```

Error : 

```
xl_more_art-full_v1.safetensors: 100%
 719M/719M [00:10<00:00, 45.4MB/s]
The current API is supported for operating with a single LoRA file. You are trying to load and fuse more than one LoRA which is not well-supported.
johnslegersart-arabscifi.0.2.safetensors: 100%
 228M/228M [00:01<00:00, 239MB/s]
The current API is supported for operating with a single LoRA file. You are trying to load and fuse more than one LoRA which is not well-supported.
```

**Comment by sayakpaul** [comment]
You should upgrade to `diffusers` latest and also install `peft`. All of that is well supported. Check out this guide for more details: https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference

**Comment by jslegers** [comment]
> You should upgrade to `diffusers` latest and also install `peft`.

I am using the latest version of diffusions




 All of that is well supported. Check out this guide for more details: https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference

**Comment by sayakpaul** [comment]
Please open a new issue with a fully reproducible code snippet then.

**Comment by FurkanGozukara** [comment]
@sayakpaul i have a question

lets say we fused lora weights

like this

```
            if lora_model_dropdown:  # Check if any LoRA models are selected
                for lora_model in lora_model_dropdown:
                    lora_path = os.path.join(used_lora_path, lora_model)
                    print(f"Loading LoRA: {lora_path}")
                    pipe.load_lora_weights(lora_path)
            
                pipe.fuse_lora(lora_scale=lora_scale_variable)
                print("LoRA models loaded and fused successfully.")
```
                
 after this if we use `pipe.unload_lora_weights() ` would it return back to base model weights? or we have to reload base model?
 
 thank you
 
 using diffusers 0.30.0 
 
 also if user changes lora scale how can i apply it without reloading entire model or lora? what would be way?
 
 what would be appropriate way of applying lora to a
```
        pipeline = StableDiffusionPipeline.from_pretrained(
            model_path,
            torch_dtype=dtypeQuantize,
        )
```

i use code like below


```
        pipeline = StableDiffusionPipeline.from_pretrained(
            model_path,
            torch_dtype=dtypeQuantize,
        )

        # Access the UNet model from the pipeline
        unet = pipeline.unet
        # Load vae
        vae = AutoencoderKL.from_pretrained(
                "madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16
        )
        print(f"vae dtype {vae.dtype}")
        pipe = StableDiffusionXLInstantIDPipeline.from_pretrained(            
            model_path,
            vae = vae,
            unet = unet,
            controlnet=[controlnet],
            torch_dtype=dtype,
            safety_checker=None,
            feature_extractor=None,
        )
```

**Comment by sayakpaul** [comment]
Follow https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference and https://huggingface.co/docs/diffusers/main/en/using-diffusers/merge_loras for detailed guidance.
